src/frontend/js/modules/avatar.js

```diff
@@ -574,6 +574,91 @@
       function displayUserStatus() {
         return self.objectType === 'user' ? !!self.avatar.id && !self.hideUserStatus : false;
       }
+    })
+    .factory('EsnInitialsAvatarGeneratorService', function() {
+      return {
+        generate
+      };
+
+      /**
+       * generate an avatar from username or email
+       *
+       * @param {String} username - the username or email
+       * @returns {String} - the generated avatar string
+       */
+      function generate(username) {
+        const text = generateInitials(username);
+        const color = generateColor(username);
+
+        return drawAvatar(text, color);
+      }
+
+      /**
+       * generate initials from username or email
+       *
+       * @param {String} name - the username or email
+       * @returns {String} - the initials
+       */
+      function generateInitials(name) {
+        const initials = name.split(' ').map(c => c[0].toUpperCase());
+
+        return initials.join('');
+      }
+
+      /**
+       * generate color from username or email
+       *
+       * @param {String} name - the username or email
+       * @returns {String} - the color
+       */
+      function generateColor(seed) {
+        let hash = 0;
+
+        seed.split('').forEach(c => {
+          // eslint-disable-next-line no-bitwise
+          hash = ((hash << 5) - hash) + c.charCodeAt(0);
```

Comment:
I don't get this part.

Reply:
creates a unique hash for a string, the hash will be the same for that string, I chose this method to avoid random colors for the same name.

Reply:
Why shift it by 5, and not 8, or 16? Why substract it after?

Reply:
detailed explanation:
https://www.javamex.com/tutorials/collections/hash_function_technical_2.shtml

tldr;
`hash << 5` is the same as `hash * 32` and `((hash << 5) - hash)` is the same as `(hash * 31)`
the reason behind using multiplication by `31` is because it's faster.

> The value 31 was chosen because it is an odd prime. If it were even and the multiplication overflowed, information would be lost, as multiplication by 2 is equivalent to shifting. The advantage of using a prime is less clear, but it is traditional. A nice property of 31 is that the multiplication can be replaced by a shift and a subtraction for better performance: 31 * i == (i << 5) - i. Modern VMs do this sort of optimization automatically.

(from Chapter 3, Item 9: Always override hashcode when you override equals, page 48)